tools/trace2syz/trace2syz/vmaCallHandlers.go

```diff
@@ -0,0 +1,357 @@
+	 */
+
+	shmid := uint64(0)
+	var fd prog.Arg
+
+	call := &prog.Call{
+		Meta: shmat,
+		Ret:  prog.MakeReturnArg(shmat.Ret),
+	}
+	ctx.CurrentSyzCall = call
+
+	if arg := ctx.ReturnCache.get(shmat.Args[0], syscall.Args[0]); arg != nil {
+		switch a := arg.Type().(type) {
+		case *prog.ResourceType:
+			fd = prog.MakeResultArg(shmat.Args[0], arg.(*prog.ResultArg), a.Default())
+		default:
+			log.Fatalf("Expected first argument of Shmat to be resource type. Got: %s\n", a.Name())
+		}
+	} else {
+		switch a := syscall.Args[0].(type) {
+		case *expression:
+			shmid = a.Eval(ctx.Target)
+		default:
+			shmid = 0
+		}
+		fd = prog.MakeResultArg(shmat.Args[0], nil, shmid)
+	}
+
+	addrArg, address := ParseAddr(pageSize, shmat.Args[1], syscall.Args[1], ctx)
+	flags := ParseFlags(shmat.Args[2], syscall.Args[2], ctx, false)
+
+	call.Args = []prog.Arg{
+		fd,
+		addrArg,
+		flags,
+	}
+	//Cache the mapped address since it is a resource type as well
+	call.Ret = prog.MakeReturnArg(shmat.Ret)
+	straceRet := newExpression(newIntType(syscall.Ret))
+	ctx.ReturnCache.cache(call.Ret.Type(), straceRet, call.Ret)
+
+	length := uint64(4096)
+	if req := ctx.State.Tracker.findShmRequest(shmid); req != nil {
+		length = req.getSize()
+	}
+	ctx.State.Tracker.createMapping(call, len(ctx.Prog.Calls), call.Args[1], address, address+length)
+	return call
+}
+
+func ParseAddr(length uint64, syzType prog.Type, traceType irType, ctx *Context) (prog.Arg, uint64) {
+	defAddrStart := (ctx.Target.NumPages - 2) * ctx.Target.PageSize
+	switch a := traceType.(type) {
+	case *pointerType:
+		var addrStart uint64
+		if a.IsNull() {
+			//Anonymous MMAP
+			addrStart = uint64(ctx.CurrentStraceCall.Ret)
+			return prog.MakeVmaPointerArg(syzType, defAddrStart, length), addrStart
+		}
+		return prog.MakeVmaPointerArg(syzType, defAddrStart, length), a.Address
+	case *expression:
+		addrStart := a.Eval(ctx.Target)
+		return prog.MakeVmaPointerArg(syzType, defAddrStart, length), addrStart
+	default:
+		panic("Failed to parse mmap")
+	}
+}
+
+func AddDependency(start, length uint64, addr prog.Arg, ctx *Context) {
+	if mapping := ctx.State.Tracker.findLatestOverlappingVMA(start); mapping != nil {
+		dependsOn := make(map[*prog.Call]int)
+		dependsOn[mapping.getCall()] = mapping.getCallIdx()
+		for _, dep := range mapping.getUsedBy() {
+			dependsOn[ctx.Prog.Calls[dep.Callidx]] = dep.Callidx
+		}
+		ctx.DependsOn[ctx.CurrentSyzCall] = dependsOn
+		dep := newMemDependency(len(ctx.Prog.Calls), addr, start, start+length)
+		mapping.addDependency(dep)
+	}
+
+}
+
+func ParseLength(straceType irType, ctx *Context) uint64 {
+	switch a := straceType.(type) {
+	case *expression:
+		return a.Eval(ctx.Target)
+	default:
+		panic("Parsing Mmap length but type is not expression")
+	}
+}
+
+func ParseFlags(syzType prog.Type, straceType irType, ctx *Context, mapFlag bool) prog.Arg {
+	switch a := straceType.(type) {
+	case *expression:
+		if mapFlag {
+			val := a.Eval(ctx.Target) | GetFixedFlag(ctx)
```

Comment:
Instead of adding FIXED flags we ought to call target.SanitizeCall for each generated call.